### ISSUE TITLE
Lazy or Non Evaluation of Block Contents

### ISSUE THREAD

**Issue by jonhartmann** [janl/mustache.js#374]
Ran into an issue attempting to use MustacheJS parser where if partial nesting statements create a recursive loop even when the data doesn't support that loop. For example, the following works:

http://jsfiddle.net/5sqdB/

The Row contains a Column which contains Text. The problem is in my data set, a Column can also contain a Row (although it doesn't in the data given). When I attempt to add the #is_row block to the #content block of the Column partial, the whole thing blows up:

http://jsfiddle.net/5sqdB/1/

It looks like the token parsing system attempts to completely parse the template before taking into consideration the actual data (since based on the data, there is no infinite loop). Could we change the system to actually pay attention if a blocks contents are necessary before attempting to request them? I've gotten this to work by refactoring it like in this post:

http://stackoverflow.com/questions/9162977/is-it-possible-to-switch-template-partial-based-on-hash-value-in-mustache

Which seems to work because the lambda is generating specific partials to include, rather than containing a list of all the partials it could include, but this solution does not work for Java based parsers (at least not the 2 I've found), and they are not open to changing their solution to support that syntax (specifically, when evaluating a lambda, they are not giving the lambda the current object being evaluated, so the lambda can't choose partials like it does in the JS version).

Basically, it would solve my problem, and would seem to be a more efficient way to handle branching includes (even if recursion wasn't an issue, if there are 50 different partial templates, the system would be requesting and evaluating all of them even though only 1 was used).

**Comment by Romanx** [comment]
I know this is old but I've been looking at lazy block evaluation for my .Net Mustache implementation and after having a look at your two fiddles It actually seems like the issue isn't with the lazy evaluation but the fact that when looking for values in the hash it will actually seek upwards in the parent context's for the value. In the case where it's getting a stack overflow error it's seeking upwards into the parent context and finding `is_row` causing infinite recursion.

Now this is the default behavior quite deliberately however maybe it's worth providing a flag of some-kind to only look within the current context when instantiating a mustache instance and default this to its current behavior?

**Comment by jonhartmann** [comment]
I think both fixes are necessary -  I was able to create an infinite loop problem by means of partial inclusion alone (even when the data did not support that execution path) and I've seen the behavior you mentioned cause duplication even when the partials did not create exact loops. I did't bring this back to this issue though as I was seeing differences between the JavaScript and Java implementations and I wasn't straight on which one was exhibiting which behavior.

**Comment by Romanx** [comment]
Is it possible for you to post an example of the infinite loop using just
partials as I want to be sure I cover off that case in my implementation

**Comment by dasilvacontin** [comment]
Good catch @Romanx.

I haven't been able to reproduce an infinite loop (for unsupported execution paths). Mind sharing, @jonhartmann?

**Comment by dasilvacontin** [comment]
Solution for the key lookup in the second fiddle: http://jsfiddle.net/5sqdB/2/.

**Comment by jonhartmann** [comment]
Weird - I thought I'd had a JSFiddle example where this wasn't working, but
couldn't locate it. Instead I recreated with a Mustache CDN reference I'd
been using at the time and got to this:

http://jsfiddle.net/jonhartmann/c2o46xk6/1/

Which is clearly not showing a problem with recursion. I'd hate to have
sent people on a goose chase - I have a clear memory of having stepped
through the execution of Mustache to see that the template was being built
before evaluation...

Apologies if this was a non-issue and I've gone off the deep end.

Thank you,

Jon Hartmann

**Comment by dasilvacontin** [comment]
No worries! We'll be glad to have you back in case you find some other issue. :)
